gcsio/src/main/java/com/google/cloud/hadoop/gcsio/ZeroCopyMessageMarshaller.java

```diff
@@ -0,0 +1,130 @@
+import io.grpc.MethodDescriptor;
+import java.io.InputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+
+// Custom gRPC marshaller to use zero memory copy feature of gRPC when deserializing messages.
+// This achieves zero-copy by deserializing proto messages pointing to the buffers in the input
+// stream to avoid memory copy so stream should live as long as the message can be referenced.
+// Hence, it exposes the input stream to applications (through popStream) and applications are
+// responsible to close it when it's no longer needed. Otherwise, it'd cause memory leak.
+class ZeroCopyMessageMarshaller<T extends MessageLite> implements PrototypeMarshaller<T> {
+  private Map<T, InputStream> unclosedStreams = Collections.synchronizedMap(new IdentityHashMap<>());
+  private final Parser<T> parser;
+  private final PrototypeMarshaller<T> baseMarshaller;
+
+  ZeroCopyMessageMarshaller(T defaultInstance) {
+    parser = (Parser<T>) defaultInstance.getParserForType();
+    baseMarshaller = (PrototypeMarshaller<T>) ProtoLiteUtils.marshaller(defaultInstance);
+  }
+
+  @Override
+  public Class<T> getMessageClass() {
+    return baseMarshaller.getMessageClass();
+  }
+
+  @Override
+  public T getMessagePrototype() {
+    return baseMarshaller.getMessagePrototype();
+  }
+
+  @Override
+  public InputStream stream(T value) {
+    return baseMarshaller.stream(value);
+  }
+
+  @Override
+  public T parse(InputStream stream) {
+    CodedInputStream cis = null;
+    try {
+      if (stream instanceof KnownLength) {
+        int size = stream.available();
+        if (stream instanceof Detachable && stream instanceof HasByteBuffer
+            && ((HasByteBuffer) stream).byteBufferSupported()) {
+          // Stream is now detached here and should be closed later.
+          stream = ((Detachable) stream).detach();
+          // This mark call is to keep buffer while traversing buffers using skip.
+          stream.mark(size);
+          List<ByteString> byteStrings = new ArrayList<>();
+          while (stream.available() != 0) {
+            ByteBuffer buffer = ((HasByteBuffer) stream).getByteBuffer();
+            byteStrings.add(UnsafeByteOperations.unsafeWrap(buffer));
+            stream.skip(buffer.remaining());
+          }
+          stream.reset();
+          cis = ByteString.copyFrom(byteStrings).newCodedInput();
+          cis.enableAliasing(true);
+          cis.setSizeLimit(Integer.MAX_VALUE);
+        }
+      }
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    if (cis != null) {
+      // fast path (no memory copy)
+      T message;
+      try {
+        message = parseFrom(cis);
+      } catch (InvalidProtocolBufferException ipbe) {
+        throw Status.INTERNAL.withDescription("Invalid protobuf byte sequence").withCause(ipbe).asRuntimeException();
```

Comment:
Is this formatted with Java Google Format?
https://github.com/google/google-java-format/blob/master/scripts/google-java-format-diff.py

Reply:
Oh I didn't know there was a tool. Thanks!